apis/buttons/src/lib.rs

```diff
@@ -0,0 +1,136 @@
+#![no_std]
+
+use libtock_platform::{
+    share::Handle, subscribe::OneId, DefaultConfig, ErrorCode, Subscribe, Syscalls, Upcall,
+};
+
+/// The Buttons driver
+///
+/// # Example
+/// ```ignore
+/// use libtock2::Buttons;
+///
+/// // Read button state
+/// Buttons::is_pressed(0);
+///
+/// // Register for events
+///
+/// let listener = ButtonListener(|button, state| {
+///     // make use of the button's state
+/// });
+///
+/// share::scope(|subscribe| {
+///     if let Ok(()) = Buttons::register_listener(&listener, subscribe) {
+///         // yield
+///     }
+/// });
+/// ```
+pub struct Buttons<S: Syscalls>(S);
+
+#[derive(Debug, PartialEq)]
+pub enum ButtonState {
+    Pressed,
+    Released,
+    Unknown,
+}
+
+impl From<u32> for ButtonState {
+    fn from(original: u32) -> ButtonState {
+        match original {
+            0 => ButtonState::Released,
+            1 => ButtonState::Pressed,
+            _ => ButtonState::Unknown,
+        }
+    }
+}
+
+impl<S: Syscalls> Buttons<S> {
+    /// Run a check against the buttons capsule to ensure it is present.
+    ///
+    /// Returns `Ok(number_of_buttons)` if the driver was present. This does not necessarily mean
+    /// that the driver is working.
+    pub fn count() -> Result<u32, ErrorCode> {
+        S::command(DRIVER_ID, BUTTONS_COUNT, 0, 0).to_result()
+    }
+
+    /// Read the state of a button
+    pub fn read(button: u32) -> Result<ButtonState, ErrorCode> {
+        let button_state: u32 = S::command(DRIVER_ID, BUTTONS_READ, button, 0).to_result()?;
+        Ok(button_state.into())
```

Comment:
Semantically, receiving a value other than `0` or `1` is an error condition. Therefore, I'm not sure that it makes sense to translate it into an `Unknown` variant. Semantically, it makes sense to translate it into `Err(ErrorCode::Fail)`, but I'd rather use code size to determine how we should handle an unexpected value.

If I compile the following snippet as a public library function, the `From` impl is 20 bytes on Thumb and 18 bytes on RISC-V:

```rust
pub enum ButtonEvent {
    Pressed,
    Released,
    Unknown,
}
impl From<u32> for ButtonEvent {
    fn from(raw: u32) -> ButtonEvent {
        match raw {
            0 => ButtonEvent::Released,
            1 => ButtonEvent::Pressed,
            _ => ButtonEvent::Unknown,
        }
    }
}
```

Note that if I redefine `ButtonEvent` so that `Released` is `0` and `Pressed` is `1`, the function shrinks to 18 bytes on Thumb but grows to 22 bytes on RISC-V.

If I try to convert unexpected values into `Err(ErrorCode::Fail)`, it grows to 34 bytes on Thumb and 30 bytes on RISC-V:

```rust
pub enum ButtonEvent {
    Pressed = 1,
    Released = 0,
}
pub fn convert(value: u32) -> Result<ButtonEvent, ErrorCode> {
    match value {
        0 => Ok(ButtonEvent::Released),
        1 => Ok(ButtonEvent::Pressed),
        _ => Err(ErrorCode::Fail),
    }
}
```

Ouch! Instead, if I just convert all unexpected values into `Pressed`, then it shrinks to 8 bytes on Thumb and 6 on RISC-V:

```rust
pub enum ButtonEvent {
    Pressed,
    Released,
}
pub fn convert(value: u32) -> ButtonEvent {
    match value {
        0 => ButtonEvent::Released,
        _ => ButtonEvent::Pressed,
    }
}
```

Therefore, I suggest removing `ButtonEvent::Unknown` and instead converting unexpected values into `ButtonEvent::Pressed`.

Reply:
Done. 

If you don't mind my asking, what command did you use to get the size of the `From` impl? Can you point me towards the documentation.

Reply:
Thank you